File: ios/RNMBX/ShapeAnimators/RNMBXChangeLineOffsetsShapeAnimatorModule.swift
import MapboxMaps
import Turf

private let LOG_TAG = "ChangeLineOffsetsShapeAnimator"

@objc
public class ChangeLineOffsetsShapeAnimator: ShapeAnimatorCommon {
  private var lineString: LineString
  private var startOfLine: AnimatableElement<Double>
  private var endOfLine: AnimatableElement<Double>
  
  init(tag: Int, lineString: LineString, startOffset: Double, endOffset: Double) {
    self.lineString = lineString
    self.startOfLine = AnimatableElement<Double>(
      source: startOffset,
      progress: startOffset,
      target: startOffset,
      startedAtSec: 0,
      progressDurationSec: 0,
      totalDurationSec: 0,
      getDistanceRemaining: { a, b in b - a }
    )
    self.endOfLine = AnimatableElement<Double>(
      source: endOffset,
      progress: endOffset,
      target: endOffset,
      startedAtSec: 0,
      progressDurationSec: 0,
      totalDurationSec: 0,
      getDistanceRemaining: { a, b in b - a }
    )
    
    super.init(tag: tag)
  }
  
  override func getShape() -> GeoJSONObject {
    return .geometry(.lineString(lineString))
  }
  
  override func getAnimatedShape(animatorAgeSec: TimeInterval) -> GeoJSONObject {
    if (startOfLine.durationRatio() < 1) {
      startOfLine.setProgress(
        value: startOfLine.source + (startOfLine.distanceRemaining() * startOfLine.durationRatio()),
        animatorAgeSec: animatorAgeSec
      )
    }
    
    if (endOfLine.durationRatio() < 1) {
      endOfLine.setProgress(
        value: endOfLine.source + (endOfLine.distanceRemaining() * endOfLine.durationRatio()),
        animatorAgeSec: animatorAgeSec
      )
    }
    
    if (startOfLine.durationRatio() >= 1 && endOfLine.durationRatio() >= 1) {
      stop()
    }
    
    if (lineString.coordinates.count < 2) {
      return emptyGeoJsonObj
    }
    
    guard let totalDistance = lineString.distance(), totalDistance > 0 else {
      return emptyGeoJsonObj
    }
    
    if (startOfLine.progress + endOfLine.progress >= totalDistance) {
      return emptyGeoJsonObj
    }
    
    guard let trimmed = lineString.trimmed(from: startOfLine.progress, to: totalDistance - endOfLine.progress) else {
      return emptyGeoJsonObj
    }
    
    return .geometry(.lineString(trimmed))
  }
  
  private func setLineString(lineString: LineString, startOffset: Double?, endOffset: Double?) {
    self.lineString = lineString
    if let _startOffset = startOffset {
      startOfLine.reset(
        _source: _startOffset,
        _progress: _startOffset,
        _target: _startOffset,
        durationSec: 0,
        animatorAgeSec: getAnimatorAgeSec()
      )
    }
    if let _endOffset = endOffset {
      endOfLine.reset(
        _source: _endOffset,
        _progress: _endOffset,
        _target: _endOffset,
        durationSec: 0,
        animatorAgeSec: getAnimatorAgeSec()
      )
    }
    refresh()
  }
  
  private func setStartOffset(offset: Double, durationSec: TimeInterval) {
    if durationSec == 0 {
      startOfLine.reset(
        _source: offset,
        _progress: offset,
        _target: offset,
        durationSec: durationSec,
        animatorAgeSec: getAnimatorAgeSec()
      )
      refresh()
    } else {
      start()
      startOfLine.reset(
        _source: startOfLine.progress,
        _progress: startOfLine.progress,
        _target: offset,
        durationSec: durationSec,
        animatorAgeSec: getAnimatorAgeSec()
      )
    }
  }
  
  private func setEndOffset(offset: Double, durationSec: TimeInterval) {
    if durationSec == 0 {
      endOfLine.reset(
        _source: offset,
        _progress: offset,
        _target: offset,
        durationSec: durationSec,
        animatorAgeSec: getAnimatorAgeSec()
      )
      refresh()
    } else {
      start()
      endOfLine.reset(
        _source: endOfLine.progress,
        _progress: endOfLine.progress,
        _target: offset,
        durationSec: durationSec,
        animatorAgeSec: getAnimatorAgeSec()
      )
    }
  }
  
  private static func getAnimator(tag: NSNumber) -> ChangeLineOffsetsShapeAnimator? {
    let animator = ShapeAnimatorManager.shared.get(tag: tag.intValue)
    return animator as? ChangeLineOffsetsShapeAnimator
  }
}

// MARK: Exposed functions

extension ChangeLineOffsetsShapeAnimator {
  @objc
  public func getTag() -> NSNumber {
    return NSNumber(value: tag)
  }

  @objc
  public static func create(tag: NSNumber, coordinates: NSArray, startOffset: NSNumber, endOffset: NSNumber) -> ChangeLineOffsetsShapeAnimator? {
    let lineString = buildLineString(_coordinates: coordinates)
    let animator = ChangeLineOffsetsShapeAnimator(tag: tag.intValue, lineString: lineString, startOffset: startOffset.doubleValue, endOffset: endOffset.doubleValue)
    ShapeAnimatorManager.shared.register(tag: tag.intValue, animator: animator)
    return animator
  }

  @objc
  public static func setLineString(tag: NSNumber, coordinates: NSArray, startOffset: NSNumber, endOffset: NSNumber, resolve: RCTPromiseResolveBlock, reject: @escaping (_ code: String, _ message: String, _ error: NSError) -> Void) {
    let lineString = buildLineString(_coordinates: coordinates)
    guard let animator = getAnimator(tag: tag) else {
      reject("\(LOG_TAG): setLineString", "Unable to find animator with tag \(tag)", NSError())
      return
    }
    
    let _startOffset: NSNumber? = startOffset != -1 ? startOffset : nil
    let _endOffset: NSNumber? = endOffset != -1 ? endOffset : nil
    
    animator.setLineString(lineString: lineString, startOffset: _startOffset?.doubleValue, endOffset: _endOffset?.doubleValue)
    resolve(tag)
  }
  
  @objc
  public static func setStartOffset(tag: NSNumber, offset: NSNumber, durationMs: NSNumber, resolve: RCTPromiseResolveBlock, reject: @escaping (_ code: String, _ message: String, _ error: NSError) -> Void) {
    guard let animator = getAnimator(tag: tag) else {
      reject("\(LOG_TAG): setStartOffset", "Unable to find animator with tag \(tag)", NSError())
      return
    }
    
    animator.setStartOffset(offset: offset.doubleValue, durationSec: durationMs.doubleValue / 1000)
    resolve(tag)
  }
  
  @objc
  public static func setEndOffset(tag: NSNumber, offset: NSNumber, durationMs: NSNumber, resolve: RCTPromiseResolveBlock, reject: @escaping (_ code: String, _ message: String, _ error: NSError) -> Void) {
    guard let animator = getAnimator(tag: tag) else {
      reject("\(LOG_TAG): setEndOffset", "Unable to find animator with tag \(tag)", NSError())
      return
    }
    
    animator.setEndOffset(offset: offset.doubleValue, durationSec: durationMs.doubleValue / 1000)
    resolve(tag)
  }
}

// - MARK: Utils

private func buildLineString(_coordinates: NSArray) -> LineString {
  let coordinates = _coordinates.map { coord in
    let coord = coord as! [NSNumber]
    return LocationCoordinate2D(latitude: coord[1].doubleValue, longitude: coord[0].doubleValue)
  }
  
  return .init(coordinates)
}

File: android/src/main/java/com/rnmapbox/rnmbx/shapeAnimators/RNMBXChangeLineOffsetsShapeAnimatorModule.kt
package com.rnmapbox.rnmbx.shapeAnimators

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.module.annotations.ReactModule
import com.mapbox.geojson.GeoJson
import com.mapbox.geojson.LineString
import com.mapbox.geojson.Point
import com.mapbox.turf.TurfConstants.UNIT_METERS
import com.mapbox.turf.TurfMeasurement
import com.mapbox.turf.TurfMisc
import com.rnmapbox.rnmbx.NativeRNMBXChangeLineOffsetsShapeAnimatorModuleSpec
import com.rnmapbox.rnmbx.utils.ViewRefTag

class ChangeLineOffsetsShapeAnimator(tag: Tag, _lineString: LineString, startOffset: Double, endOffset: Double): ShapeAnimatorCommon(tag) {
    private var lineString = _lineString
    private var startOfLine = AnimatableElement<Double>(
        startOffset,
        startOffset,
        startOffset,
        0.0,
        0.0,
        0.0,
        { a, b -> b - a }
    )
    private var endOfLine = AnimatableElement<Double>(
        endOffset,
        endOffset,
        endOffset,
        0.0,
        0.0,
        0.0,
        { a, b -> b - a }
    )

    override fun getAnimatedShape(animatorAgeSec: Double): GeoJson {
        if (startOfLine.durationRatio() < 1) {
            startOfLine.setProgress(
                startOfLine.source + (startOfLine.distanceRemaining() * startOfLine.durationRatio()),
                animatorAgeSec
            )
        }

        if (endOfLine.durationRatio() < 1) {
            endOfLine.setProgress(
                endOfLine.source + (endOfLine.distanceRemaining() * endOfLine.durationRatio()),
                animatorAgeSec
            )
        }

        if (startOfLine.durationRatio() >= 1 && endOfLine.durationRatio() >= 1) {
            stop()
        }

        if (lineString.coordinates().count() < 2) {
            return emptyGeoJsonObj
        }

        val totalDistance = TurfMeasurement.length(lineString, UNIT_METERS)
        if (totalDistance == 0.0) {
            return emptyGeoJsonObj
        }

        if (startOfLine.progress + endOfLine.progress >= totalDistance) {
            return emptyGeoJsonObj
        }

        val trimmed = TurfMisc.lineSliceAlong(
            lineString,
            startOfLine.progress,
            totalDistance - endOfLine.progress,
            UNIT_METERS
        )
        return trimmed
    }

    fun setLineString(lineString: LineString, startOffset: Double?, endOffset: Double?) {
        this.lineString = lineString
        if (startOffset != null) {
            startOfLine.reset(
                startOffset,
                startOffset,
                startOffset,
                0.0,
                getAnimatorAgeSec()
            )
        }
        if (endOffset != null) {
            endOfLine.reset(
                endOffset,
                endOffset,
                endOffset,
                0.0,
                getAnimatorAgeSec()
            )
        }
        refresh()
    }

    fun setStartOffset(offset: Double, durationSec: Double) {
        if (durationSec == 0.0) {
            startOfLine.reset(
                offset,
                offset,
                offset,
                durationSec,
                getAnimatorAgeSec()
            )
            refresh()
        } else {
            start()
            startOfLine.reset(
                startOfLine.progress,
                startOfLine.progress,
                offset,
                durationSec,
                getAnimatorAgeSec()
            )
        }
    }

    fun setEndOffset(offset: Double, durationSec: Double) {
        if (durationSec == 0.0) {
            endOfLine.reset(
                offset,
                offset,
                offset,
                durationSec,
                getAnimatorAgeSec()
            )
            refresh()
        } else {
            start()
            endOfLine.reset(
                endOfLine.progress,
                endOfLine.progress,
                offset,
                durationSec,
                getAnimatorAgeSec()
            )
        }
    }
}

@ReactModule(name = RNMBXChangeLineOffsetsShapeAnimatorModule.NAME)
class RNMBXChangeLineOffsetsShapeAnimatorModule(
    reactContext: ReactApplicationContext?,
    val shapeAnimatorManager: ShapeAnimatorManager
): NativeRNMBXChangeLineOffsetsShapeAnimatorModuleSpec(reactContext) {
    companion object {
        const val LOG_TAG = "RNMBXChangeLineOffsetsShapeAnimatorModule"
        const val NAME = "RNMBXChangeLineOffsetsShapeAnimatorModule"
    }

    override fun create(
        tag: ViewRefTag,
        coordinates: ReadableArray,
        startOffset: Double,
        endOffset: Double,
        promise: Promise?
    ) {
        val lineString = buildLineString(coordinates)

        shapeAnimatorManager.add(
            ChangeLineOffsetsShapeAnimator(
                tag.toLong(),
                lineString,
                startOffset,
                endOffset
            )
        )
        promise?.resolve(tag.toInt())
    }

    private fun getAnimator(tag: ViewRefTag): ChangeLineOffsetsShapeAnimator {
        return shapeAnimatorManager.get(tag.toLong()) as ChangeLineOffsetsShapeAnimator
    }

    override fun setLineString(tag: ViewRefTag, coordinates: ReadableArray?, startOffset: Double, endOffset: Double, promise: Promise?) {
        val animator = getAnimator(tag)

        if (coordinates == null) {
            return
        }

        val _startOffset = if (startOffset != -1.0) startOffset else null
        val _endOffset = if (endOffset != -1.0) endOffset else null

        val lineString = buildLineString(coordinates)
        animator.setLineString(lineString, _startOffset, _endOffset)
        promise?.resolve(true)
    }

    override fun setStartOffset(tag: ViewRefTag, offset: Double, duration: Double, promise: Promise?) {
        val animator = getAnimator(tag)
        animator.setStartOffset(offset, duration / 1000)
        promise?.resolve(true)
    }

    override fun setEndOffset(tag: ViewRefTag, offset: Double, duration: Double, promise: Promise?) {
        val animator = getAnimator(tag)
        animator.setEndOffset(offset, duration / 1000)
        promise?.resolve(true)
    }
}

private fun buildLineString(_coordinates: ReadableArray): LineString {
    var coordinates: List<Point> = listOf()

    for (i in 0 until _coordinates.size()) {
        val arr = _coordinates.getArray(i)
        val coord = Point.fromLngLat(arr.getDouble(0), arr.getDouble(1))
        coordinates = coordinates.plus(coord)
    }

    return LineString.fromLngLats(coordinates)
}